src/ring_theory/witt_vector/isocrystal.lean

```diff
@@ -0,0 +1,201 @@
+
+/-- The Frobenius automorphism of `k` induces an endomorphism of `K`. For notation purposes. -/
+def fraction_ring.frobenius_ring_hom : K(p, k) →+* K(p, k) := fraction_ring.frobenius p k
+
+localized "notation `φ(` p`,` k`)` := witt_vector.fraction_ring.frobenius_ring_hom p k"
+  in isocrystal
+
+instance inv_pair₁ : ring_hom_inv_pair (φ(p, k)) _ :=
+ring_hom_inv_pair.of_ring_equiv (fraction_ring.frobenius p k)
+
+instance inv_pair₂ :
+  ring_hom_inv_pair ((fraction_ring.frobenius p k).symm : K(p, k) →+* K(p, k)) _ :=
+ring_hom_inv_pair.of_ring_equiv (fraction_ring.frobenius p k).symm
+
+localized "notation M ` →ᶠˡ[`:50 p `,` k `] ` M₂ :=
+  linear_map (witt_vector.fraction_ring.frobenius_ring_hom p k) M M₂" in isocrystal
+localized "notation M ` ≃ᶠˡ[`:50 p `,` k `] ` M₂ :=
+  linear_equiv (witt_vector.fraction_ring.frobenius_ring_hom p k) M M₂" in isocrystal
+
+/-! ### Isocrystals -/
+
+/--
+An isocrystal is a vector space over the field `K(p, k)` additionally equipped with a
+Frobenius-linear automorphism.
+-/
+class isocrystal (V : Type*) [add_comm_group V] extends module K(p, k) V :=
+( frob : V ≃ᶠˡ[p, k] V )
+
+variables (V : Type*) [add_comm_group V] [isocrystal p k V]
+variables (V₂ : Type*) [add_comm_group V₂] [isocrystal p k V₂]
+
+variables {V}
+
+/--
+Project the Frobenius automorphism from an isocrystal. Denoted by `Φ(p, k)` when V can be inferred.
+-/
+def isocrystal.frobenius : V ≃ᶠˡ[p, k] V := @isocrystal.frob p _ k _ _ _ _ _ _ _
+variables (V)
+
+localized "notation `Φ(` p`,` k`)` := witt_vector.isocrystal.frobenius p k" in isocrystal
+
+/-- A homomorphism between isocrystals respects the Frobenius map. -/
+@[nolint has_inhabited_instance]
+structure isocrystal_hom extends V →ₗ[K(p, k)] V₂ :=
+( frob_equivariant : ∀ x : V, Φ(p, k) (to_linear_map x) = to_linear_map (Φ(p, k) x) )
+
+/-- An isomorphism between isocrystals respects the Frobenius map. -/
+@[nolint has_inhabited_instance]
+structure isocrystal_equiv extends V ≃ₗ[K(p, k)] V₂ :=
+( frob_equivariant : ∀ x : V, Φ(p, k) (to_linear_equiv x) = to_linear_equiv (Φ(p, k) x) )
+
+localized "notation M ` →ᶠⁱ[`:50 p `,` k `] ` M₂ := witt_vector.isocrystal_hom p k M M₂"
+  in isocrystal
+localized "notation M ` ≃ᶠⁱ[`:50 p `,` k `] ` M₂ := witt_vector.isocrystal_equiv p k M M₂"
+  in isocrystal
+
+
+end perfect_ring
+
+open_locale isocrystal
+
+/-! ### Classification of isocrystals in dimension 1 -/
+
+/-- A helper instance for type class inference. -/
+local attribute [instance]
+def fraction_ring.module : module K(p, k) K(p, k) := semiring.to_module
+
+/--
+Type synonym for `K(p, k)` to carry the standard `m`-indexed 1-dimensional isocrystal structure.
+-/
+@[nolint unused_arguments has_inhabited_instance, derive [add_comm_group, module K(p, k)]]
+def standard_one_dim_isocrystal (m : ℤ) : Type* :=
+K(p, k)
+
+section perfect_ring
+variables [is_domain k] [char_p k p] [perfect_ring k p]
+
+/-- The standard one-dimensional isocrystal is an isocrystal. -/
```

Comment:
```suggestion
/-- The standard one-dimensional isocrystal of weight `m : ℤ` is an isocrystal. -/
```

Reply:
Hmmm... The word `weight` might have different mathematical connotations in this context. I would choose a different word.

Reply:
Oh, my apologies, I misremembered @kbuzzard's suggestion:  it was "slope".  Do you agree with that one?

Reply:
Yes! `slope` should be the right word here.